### PR TITLE
Set default application language to I18n.locale

### PR DESCRIPTION
**Pull request by metamoni**
### Summary
This PR aims to improve the accessibility of Rails apps by adding a `lang` attribute to HTML templates generated by the framework. I started with `application.html.erb` and set its `lang` to `I18n.locale`, so the language gets automatically updated when the locale is updated. There's a lot more work to be done, as _all_ html templates need to be updated. So this would be the pilot, I suppose 🙂

Benefits include:
- making it easier for translation services to detect the page language
- helping screen reader software read things out properly
- improving the experience for people who use text-to-speech software

It's also a step towards meeting [Success Criterion 3.1.1: Language of Page](https://www.w3.org/WAI/WCAG21/Understanding/language-of-page) of the Web Content Accessibility Guidelines.

### What these changes do
Modifying the template for `application.html.erb` will have the following effect: in any new Rails app, all new pages generated via `rails g controller ControllerName action` will be created with a `lang` in their HTML. The language will then evaluate to whatever `I18n.locale` is.

To test this:
- pull down the branch
- run `bundle exec rails new ~/my-test-app --dev` to generate an app that uses the branch
- `cd ~/my-test-app` then run `bundle install`, `rails webpacker:install` to install dependencies
- generate a new controller with `rails g controller Home index`
- start the server with `rails s`
- go to `http://localhost:3000/home/index`, inspect the page and verify that the HTML tag has `lang="en"`
- in `config.application.rb`, change your default locale to German with `config.i18n.default_locale = :de`
- restart the server and navigate to `http://localhost:3000/home/index` - you should be prompted to translate the page (in Google Chrome) and when inspecting the page you should now see `html lang="de"`

### Other Information
- I would have liked to have examples that check that the rendered html includes `<html lang="en">` by default, and `<html lang="de">` when I set `I18n.locale = :de`. However, I'm not sure _where_ to include these tests. If you can think of any better tests I could write, please help point me in the right direction
- wasn't sure whether I had to update the CHANGELOG for this. Happy to remove this if needed
- read full [forum discussion](https://discuss.rubyonrails.org/t/accessibility-proposal-add-lang-attribute-to-generated-views/77634/4) here

### CODE REVIEW

File: railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <!DOCTYPE html>
-<html>
+<html lang=<%%= I18n.locale %>>
```
Comment (zzak) [inline railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:2]: Does this need to be quoted? `<html lang="<%%= I18n.locale %>">`

File: railties/test/generators/app_generator_test.rb [hunk 2]
```diff
@@ -1238,6 +1238,15 @@ def test_minimal_rails_app
     assert_no_gem "web-console", app_root
   end
 
+  def test_application_template_has_locale
+    app_root = File.join(destination_root, "myapp")
+    run_generator [app_root, "--minimal"]
+
+    assert_file "#{app_root}/app/views/layouts/application.html.erb" do |content|
+      assert_match(/I18n.locale/, content)
```
Comment (zzak) [inline railties/test/generators/app_generator_test.rb:1246]: Is this matching the string "I18n.locale"? Or should match the value, like "en"?

File: railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt (hunk 1 above)
Comment (metamoni) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:2]: I'm inserting the value of `<%= I18n.locale %>`, which evaluates to the string "en" (or the symbol `:en`), so it doesn't need to be quoted here.

File: railties/test/generators/app_generator_test.rb (hunk 2 above)
Comment (metamoni) [reply railties/test/generators/app_generator_test.rb:1246]: This test doesn't check the outputted HTML, it checks that `I18n.locale` is inserted into the `application.html.erb` template. So the template would need to include `I18n.locale` (or maybe the full `lang=<%%= I18n.locale %>`), and in the rendered HTML that would evaluate to `lang="en"`. I'm not sure where to add a test that checks the rendered HTML, but this spec is only for the template.

File: railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt (hunk 1 above)
Comment (metamoni) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:2]: @zzak I've updated the PR with some more context about what these changes are, what they do in the app, and some QA steps

File: railties/test/generators/app_generator_test.rb (hunk 2 above)
Comment (zzak) [reply railties/test/generators/app_generator_test.rb:1246]: @metamoni Right, I was so confused but that makes sense 😅

**Comment by hahmed** [comment]
Nice @metamoni, 👍 for this change.

> I would have liked to have examples that check that the rendered html includes <html lang="en"> by default, and <html lang="de"> when I set I18n.locale = :de. However, I'm not sure where to include these tests. If you can think of any better tests I could write, please help point me in the right direction

I think you could add a test to the file `railties/test/application/initializers/i18n_test.rb`, maybe something like below?

```
test "displays correct lang in root tag" do
      require "rack/test"
      extend Rack::Test::Methods

      add_to_config <<-RUBY
        config.i18n.default_locale = :de
      RUBY

      controller :posts, <<-RUBY
        class PostsController < ActionController::Base
          def index
            render :index, layout: "application"
          end
        end
      RUBY

      app_file "config/routes.rb", <<-RUBY
        Rails.application.routes.draw do
          root to: "posts#index"
        end
      RUBY

      app_file "app/views/posts/index.html.erb", "hello"

      load_app

      get "/"
      assert_match "<html lang=\"de\">", last_response.body
    end
```

**Review by lfalcao — APPROVED** [review]

File: railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt (hunk 1 above)
Comment (metamoni) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt:2]: I'll update this

File: railties/test/generators/app_generator_test.rb [hunk 3]
```diff
@@ -1238,6 +1238,15 @@ def test_minimal_rails_app
     assert_no_gem "web-console", app_root
   end
 
+  def test_application_template_has_locale
+    app_root = File.join(destination_root, "myapp")
+    run_generator [app_root, "--minimal"]
```
[the remark this answers is not in the document]
Comment (metamoni) [reply railties/test/generators/app_generator_test.rb:1243]: Error on my part, left that in by accident. I'll remove it

File: railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt (hunk 1 above)
Comment (metamoni) [reply railties/lib/rails/generators/rails/app/templates/app/views/layouts/application.html.erb.tt]: Just to clarify, after updating this to `<html lang="<%%= I18n.locale %>">` the generated `application.html.erb` file will contain `<html lang="<% I18n.locale %>"`, whereas `<html lang=<%%= I18n.locale %>>` generates it as `<html lang=<% I18n.locale %>`. In both cases, when inspecting this in the browser it'll show as `<html lang="en">`.

**Comment by metamoni** [comment]
> Nice @metamoni, 👍 for this change.
> 
> > I would have liked to have examples that check that the rendered html includes  by default, and  when I set I18n.locale = :de. However, I'm not sure where to include these tests. If you can think of any better tests I could write, please help point me in the right direction
> 
> I think you could add a test to the file `railties/test/application/initializers/i18n_test.rb`?

Good shout, thanks! I've added a test to that file too

File: railties/test/application/initializers/i18n_test.rb [hunk 4]
```diff
@@ -45,6 +45,13 @@ def assert_no_fallbacks
       assert_equal :de, I18n.default_locale
     end
 
+    test "generating application template with locale" do
+      load_app
+      
+      application_layout = File.read("#{app_path}/app/views/layouts/application.html.erb")
+      assert_includes application_layout, "<html lang=\"<%= I18n.locale %>\">"
```
Comment (rafaelfranca) [inline railties/test/application/initializers/i18n_test.rb:52]: Isn't this testing the same as the test in app_generator_test? I think what you want is a test that renders the page, so it would be closer to this test here https://github.com/rails/rails/blob/09d55b302266cf002a4b307f8d37a105d2838a18/railties/test/application/content_security_policy_test.rb#L19-L38

Comment (metamoni) [reply railties/test/application/initializers/i18n_test.rb:52]: Hmm sorry, my test was looking at the actual `application.html.erb` file, not at the HTML output. Thanks for sharing that example, I'll write a proper test for this.

Comment (metamoni) [reply railties/test/application/initializers/i18n_test.rb:52]: @rafaelfranca before I update this, is `railties/test/application/initializers/i18n_test.rb` still the best place for this test?

Comment (hahmed) [reply railties/test/application/initializers/i18n_test.rb:52]: Yup, you can still add the test there 👍🏼

File: railties/CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,9 @@
     Fixes #40744.
 
     *Michael Duchemin*
+*   Set default language for `application.html.erb` to `I18n.locale`.
```
Comment (sunny) [inline railties/CHANGELOG.md:8]: Minor line-break suggestion here:

```suggestion

*   Set default language for `application.html.erb` to `I18n.locale`.
```

File: railties/test/application/initializers/i18n_test.rb (hunk 4 above)
Comment (metamoni) [reply railties/test/application/initializers/i18n_test.rb]: Added two new tests. Please have a look and let me know if I'm on the right track

**Comment by zzak** [comment]
@metamoni Could you check the RuboCop failures?

```
railties/test/application/initializers/i18n_test.rb:67:1: C: [Correctable] Layout/TrailingWhitespace: Trailing whitespace detected.
railties/test/application/initializers/i18n_test.rb:71:1: C: [Correctable] Layout/TrailingWhitespace: Trailing whitespace detected.
```

**Review by hahmed — APPROVED** [review]
Awesome, thanks for the PR @metamoni ✨

**Comment by matthewd** [comment]
Sorry I'm late to the party here with such a fundamental question, but: does it really improve accessibility for us to default to claiming all responses are in the configured language? 'cos they're.. not.

If anything it seems to me it's more honest to generate `lang="en"` -- though that's also problematic, because it's likely to create new applications claiming to be in English even after their actual content has been changed to some other language. (That risk is present even when using the i18n-configured locale: we have no guarantee that people will suitably change it, instead of just putting their local strings into the already-configured bucket.)

I've always seen it as a deliberate choice that we _don't_ specify a language, preferring that, if a user doesn't make the effort to set it, we'll have contributed to the "not specified; you're going to have to make a guess, sorry" corpus instead of the "claims to be [lang1], but now you'll need to infer that's a lie _and_ guess that the content is actually [lang2]" wasteland.

In other words, I agree with the goal of more pages/applications having correct language attributes... but I think it's worth exploring whether we as a framework are in a position to make that happen, vs the risk that we contribute to contaminating the pool of correctly-tagged content. 

I'd like to consider this change (claiming the configured i18n language) against generating a hard-coded "en", perhaps against generating a commented version of one of those two. I imagine you might have given this balance some consideration already, so I'd love to hear your thoughts. Is there precedent for this in other similarly removed-from-the-actual-content frameworks?

**Comment by metamoni** [comment]
> Sorry I'm late to the party here with such a fundamental question, but: does it really improve accessibility for us to default to claiming all responses are in the configured language? 'cos they're.. not.
> 
> If anything it seems to me it's more honest to generate `lang="en"` -- though that's also problematic, because it's likely to create new applications claiming to be in English even after their actual content has been changed to some other language. (That risk is present even when using the i18n-configured locale: we have no guarantee that people will suitably change it, instead of just putting their local strings into the already-configured bucket.)
> 
> I've always seen it as a deliberate choice that we _don't_ specify a language, preferring that, if a user doesn't make the effort to set it, we'll have contributed to the "not specified; you're going to have to make a guess, sorry" corpus instead of the "claims to be [lang1], but now you'll need to infer that's a lie _and_ guess that the content is actually [lang2]" wasteland.
> 
> In other words, I agree with the goal of more pages/applications having correct language attributes... but I think it's worth exploring whether we as a framework are in a position to make that happen, vs the risk that we contribute to contaminating the pool of correctly-tagged content.
> 
> I'd like to consider this change (claiming the configured i18n language) against generating a hard-coded "en", perhaps against generating a commented version of one of those two. I imagine you might have given this balance some consideration already, so I'd love to hear your thoughts. Is there precedent for this in other similarly removed-from-the-actual-content frameworks?

Seems to me like the framework has no issues [defining the language in some places](https://github.com/rails/rails/search?q=html+lang%3D%22en%22). At the moment, however, this behaviour is inconsistent across various templates, so all I'm trying to do here is start a pattern where templates have a default language and then consistently get updated when developers set another default locale.
